decision-records/2024-10-09--adding-k8s-labels-from-libs.md

```diff
@@ -0,0 +1,56 @@
+# Adding k8s Labels From Libraries
+**Date:** 2024-07-16  
+**Authors:** @dstathis @ca-scribner
+
+## Context and Problem Statement
+
+In order for charms to join the mesh, beacon sends a set of k8s labels over relation data that the charm must apply to
+its pods. The `service_mesh` library should attempt to manage those labels.
+
+The following must be true:
+
+1. Labels must match relation data after a leader change
+2. Labels must match relation data after a pod restart
+3. Labels must match relation data when relation data changes
+4. Labels applied by anything outside the library should remain untouched
+
+## Original (Bad) Solution:
+
+* Only the leader processes labels
+* Whenever the leader makes a change,  put the dict of labels it is responsible for in a StoredState.
+* When a hook fires compare relation data to the stored state and apply the diff
+
+## The Problem:
+
+The issue seems to be when 1 and 3 happen at the same time (or 2 and 3). We will have no idea how the labels have
+changed. For example, lets say relation data is `{"foo": "bar"}` and labels are set correctly. Then the relation data is
+set reset to `{}` but the the leader pod dies at the same time. The new leader will come up and process the relation
+changed event. It will see that it has no stored state and that matches relation data. So it will do nothing. The
+`{"foo": "bar"}` label is now left on the pods unmanaged.
+
+The issue stems from the fact that StoredState is scoped to the pod and not the charm. Even if `use_juju_for_storage` is
+set to True, this issue will still persist. The usual solution would be to use a peer relation instead of StoredState
+but since this is a library, we can't really impose this on the charm (or can we?)
```

Comment:
we could. Make it a mandatory arg in the requirer constructor and force the charm to specify a peer relation we can use for persisting data. I think there's a precedent for this but I can't remember where.
It is however one extra step for the charm owner.

Reply:
another (likely bad) solution could be to create a secret and use it for persistent storage :D

Reply:
We considered the first solution and decided it is better not to require a peer relation. The second idea is basically the same as the accepted solution. The data isn't secret so we use a config map instead.